packages/firestore/src/model/transform_operation.ts

```diff
@@ -186,35 +183,32 @@ export class ArrayRemoveTransformOperation implements TransformOperation {
  * arithmetic is used and precision loss can occur for values greater than 2^53.
  */
 export class NumericIncrementTransformOperation implements TransformOperation {
-  constructor(readonly operand: NumberValue) {}
+  constructor(
+    private readonly serializer: JsonProtoSerializer,
+    readonly operand: api.Value
+  ) {
+    assert(
+      isNumber(operand),
+      'NumericIncrementTransform transform requires a NumberValue'
+    );
+  }
 
   applyToLocalView(
-    previousValue: FieldValue | null,
+    previousValue: api.Value | null,
     localWriteTime: Timestamp
-  ): FieldValue {
-    const baseValue = this.computeBaseValue(previousValue);
+  ): api.Value {
     // PORTING NOTE: Since JavaScript's integer arithmetic is limited to 53 bit
     // precision and resolves overflows by reducing precision, we do not
     // manually cap overflows at 2^63.
-
-    // Return an integer value iff the previous value and the operand is an
-    // integer.
-    if (
-      baseValue instanceof IntegerValue &&
-      this.operand instanceof IntegerValue
-    ) {
-      const sum = baseValue.internalValue + this.operand.internalValue;
-      return new IntegerValue(sum);
-    } else {
-      const sum = baseValue.internalValue + this.operand.internalValue;
-      return new DoubleValue(sum);
-    }
+    const baseValue = this.computeBaseValue(previousValue);
+    const sum = this.asNumber(baseValue) + this.asNumber(this.operand);
+    return this.serializer.toNumber(sum);
```

Comment:
Does this give the same result for all possible integer values? It seems like if you add to a sum that's larger than `isSafeInteger(sum)` this will now return a `doubleValue` where it used to be an `integerValue`. Though the calculation lost precision, I'm not sure changing the type is appropriate.